File: common/src/main/java/pers/fz/mvvm/base/BaseFragment.java
package pers.fz.mvvm.base;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import org.jetbrains.annotations.NotNull;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import pers.fz.mvvm.inter.ErrorService;
import pers.fz.mvvm.util.permission.PermissionsChecker;
import pers.fz.mvvm.wight.dialog.LoadingProgressDialog;

/**
 * Created by fz on 2017/11/22.
 * BaseFragment封装
 */
public abstract class BaseFragment<VM extends BaseViewModel, VDB extends ViewDataBinding> extends Fragment implements BaseView {
    protected String TAG = this.getClass().getSimpleName();

    protected VM mViewModel;
    protected VDB binding;
    private ActivityResultLauncher<String[]> permissionLauncher;
    @Inject
    protected ErrorService errorService;
    protected ActivityResultLauncher<Intent> loginLauncher = null;

    public void requestPermission(String... permissions) {
        permissionLauncher.launch(permissions);
    }

    public void requestPermission(List<String> permissions) {
        // 缺少权限时, 进入权限配置页面
        permissionLauncher.launch(permissions.stream().toArray(String[]::new));
    }

    public boolean lacksPermissions(String... permissions) {
        return PermissionsChecker.getInstance().lacksPermissions(requireContext(), permissions);
    }

    public boolean lacksPermissions(List<String> permission) {
        return PermissionsChecker.getInstance().lacksPermissions(requireContext(), permission);
    }

    /**
     * 注册权限请求监听
     */
    protected void registerPermissionLauncher() {
        permissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), result -> {
            for (Map.Entry<String, Boolean> stringBooleanEntry : result.entrySet()) {
                if (Boolean.FALSE.equals(stringBooleanEntry.getValue())) {
                    onPermissionRefused(result);
                    return;
                }
            }
            onPermissionGranted(result);
        });
    }

    protected void unregisterPermission() {
        if (permissionLauncher != null) {
            permissionLauncher.unregister();
        }
    }

    /**
     * 权限同意
     */
    protected void onPermissionGranted(Map<String, Boolean> permissions) {

    }

    /**
     * 权限拒绝
     */
    protected void onPermissionRefused(Map<String, Boolean> permissions) {
        showToast("拒绝权限可能会导致应用软件运行异常!");
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);
        binding.setLifecycleOwner(this);
        loginLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
            if (result.getResultCode() == Activity.RESULT_OK) {
                onLoginSuccessCallback(result.getData() == null ? null : result.getData().getExtras());
            } else {
                onLoginFailCallback(result.getResultCode(), result.getData() == null ? null : result.getData().getExtras());
            }
        });
        createViewModel();
        initView(savedInstanceState);
        initData(getArguments() == null ? new Bundle() : getArguments());
        return binding.getRoot();
    }

    /**
     * 创建viewModel
     */
    public void createViewModel() {
        if (mViewModel == null) {
            Class modelClass;
            Type type = getClass().getGenericSuperclass();
            if (type instanceof ParameterizedType) {
                modelClass = (Class) ((ParameterizedType) type).getActualTypeArguments()[0];
            } else {
                //如果没有指定泛型参数，则默认使用BaseViewModel
                modelClass = BaseViewModel.class;
            }
            mViewModel = (VM) new ViewModelProvider(useActivityViewModel() ? requireActivity() : this).get(modelClass);
            mViewModel.createRepository(useActivityViewModel() ? ((BaseActivity)requireActivity()) : this);
        }
    }

    /**
     * 是否和activity共用同一个viewModel
     * @return true代表与Activity共用
     */
    public boolean useActivityViewModel() {
        return false;
    }

    /**
     * 该抽象方法就是 onCreateView中需要的layoutID
     *
     * @return 布局资源id
     */
    protected abstract int getLayoutId();

    /**
     * 该抽象方法就是 初始化view
     */
    protected abstract void initView(Bundle savedInstanceState);

    /**
     * 执行数据的加载
     */
    protected abstract void initData(Bundle bundle);

    protected void onLoginSuccessCallback(Bundle bundle) {

    }

    protected void onLoginFailCallback(int resultCode, Bundle bundle) {

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (loginLauncher != null) {
            loginLauncher.unregister();
        }
    }

    /**
     * 显示加载弹框
     *
     * @param dialogMessage 弹框内容，如果内容为空则不展示文字部分
     */
    private void showLoadingDialog(String dialogMessage, boolean isCanCancel) {
        LoadingProgressDialog.getInstance(getContext())
                .setCanCancel(isCanCancel)
                .setMessage(dialogMessage)
                .builder()
                .show();
    }

    @Override
    public void showLoading(String dialogMessage) {
        if (!requireActivity().isFinishing()) {
            requireActivity().runOnUiThread(() -> showLoadingDialog(dialogMessage, false));
        }
    }

    @Override
    public void hideLoading() {
        if (!requireActivity().isFinishing()) {
            requireActivity().runOnUiThread(() -> LoadingProgressDialog.getInstance(getContext()).dismiss());
        }
    }

    @Override
    public void refreshLoading(String dialogMessage) {
        if (!requireActivity().isFinishing()) {
            requireActivity().runOnUiThread(() -> {
                if (LoadingProgressDialog.getInstance(requireActivity()) == null || !LoadingProgressDialog.getInstance(requireActivity()).isShowing()) {
                    return;
                }
                LoadingProgressDialog.getInstance(requireActivity())
                        .refreshMessage(dialogMessage);
            });
        }
    }

    @Override
    public void showToast(String msg) {
        if (TextUtils.isEmpty(msg)) {
            return;
        }
        requireActivity().runOnUiThread(() -> Toast.makeText(requireActivity(), msg, Toast.LENGTH_SHORT).show());
    }

    @Override
    public void onErrorCode(BaseModelEntity model) {
        if (errorService == null || model == null) {
            return;
        }
        if (!errorService.isLoginPast(model.getCode())) {
            errorService.toLogin(requireContext(), loginLauncher);
            return;
        }
        if (!errorService.hasPermission(model.getCode())) {
            errorService.toNoPermission(requireContext());
        }
    }

    public void startActivity(Class<?> toClx) {
        startActivity(toClx, null);
    }

    public void startActivity(Class<?> toClx, Bundle bundle) {
        Intent intent = new Intent(requireContext(), toClx);
        if (bundle != null) {
            intent.putExtras(bundle);
        }
        startActivity(intent);
    }

    public void startForResult(ActivityResultLauncher<Intent> activityResultLauncher, Class<?> toClx) {
        startForResult(activityResultLauncher, toClx, null);
    }

    public void startForResult(ActivityResultLauncher<Intent> activityResultLauncher, Class<?> toClx, Bundle bundle) {
        Intent intent = new Intent(requireContext(), toClx);
        if (bundle != null) {
            intent.putExtras(bundle);
        }
        activityResultLauncher.launch(intent);
    }

    public void startForResult(ActivityResultLauncher<Intent> activityResultLauncher, Intent intent) {
        activityResultLauncher.launch(intent);
    }

}

File: common/src/main/java/pers/fz/mvvm/base/kotlin/BaseFragment.kt
package pers.fz.mvvm.base.kotlin

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import pers.fz.mvvm.base.BaseModelEntity
import pers.fz.mvvm.base.BaseView
import pers.fz.mvvm.inter.ErrorService
import pers.fz.mvvm.util.log.ToastUtils
import pers.fz.mvvm.util.permission.PermissionsChecker
import pers.fz.mvvm.wight.dialog.CustomProgressDialog
import java.lang.reflect.ParameterizedType
import javax.inject.Inject

/**
 * Created by fz on 2017/11/22.
 * BaseFragment封装
 */
abstract class BaseFragment<VM : BaseViewModel<V>?, VDB : ViewDataBinding?, V : BaseView> :
    Fragment(), BaseView {
    protected var TAG = this.javaClass.simpleName

    /**
     * 权限检测器
     */
    private var mPermissionsChecker: PermissionsChecker? = null
    protected var mViewModel: VM? = null
    protected var binding: VDB? = null
    private var permissionLauncher: ActivityResultLauncher<Array<String>>? = null

    @Inject
    lateinit var errorService: ErrorService
    protected var loginLauncher: ActivityResultLauncher<Intent>? = null
    fun requestPermission(permissions: Array<String>) {
        permissionLauncher!!.launch(permissions)
    }

    private val permissionsChecker: PermissionsChecker
        get() {
            if (mPermissionsChecker == null) {
                mPermissionsChecker = PermissionsChecker(requireActivity())
            }
            return mPermissionsChecker as PermissionsChecker
        }

    fun lacksPermissions(permissions: Array<String?>): Boolean {
        return permissionsChecker.lacksPermissions(*permissions)
    }

    /**
     * 注册权限请求监听
     */
    protected fun registerPermissionLauncher() {
        permissionLauncher = registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions()
        ) { permissions ->
            run {
                val allPermissionsGranted = permissions.values.all { it }
                if (allPermissionsGranted) {
                    onPermissionGranted(permissions)
                } else {
                    onPermissionRefused(permissions)
                }
            }
        }
    }

    protected fun unregisterPermission() {
        if (permissionLauncher != null) {
            permissionLauncher!!.unregister()
        }
    }

    /**
     * 权限同意
     */
    protected fun onPermissionGranted(permissions: Map<String, Boolean>?) {}

    /**
     * 权限拒绝
     */
    protected fun onPermissionRefused(permissions: Map<String, Boolean>?) {
        showToast("拒绝权限可能会导致应用软件运行异常!")
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater, layoutId, container, false)
        binding!!.lifecycleOwner = this
        loginLauncher =
            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
                if (result.resultCode == Activity.RESULT_OK) {
                    onLoginSuccessCallback(if (result.data == null) null else result.data!!.extras)
                } else {
                    onLoginFailCallback(
                        result.resultCode, if (result.data == null) null else result.data!!
                            .extras
                    )
                }
            }
        createViewModel()
        initView(savedInstanceState)
        initData(if (arguments == null) Bundle() else arguments)
        return binding!!.root
    }

    fun createViewModel() {
        if (mViewModel == null) {
            mViewModel = ViewModelProvider(this).get(getVmClazz(this))
            mViewModel?.baseView = this as V
        }
    }

    /**
     * 获取当前类绑定的泛型ViewModel-clazz
     */
    @Suppress("UNCHECKED_CAST")
    fun <VM> getVmClazz(obj: Any): VM {
        return (obj.javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as VM
    }

    protected abstract val layoutId: Int

    /**
     * 该抽象方法就是 初始化view
     */
    protected abstract fun initView(savedInstanceState: Bundle?)

    /**
     * 执行数据的加载
     */
    protected abstract fun initData(bundle: Bundle?)
    protected open fun onLoginSuccessCallback(bundle: Bundle?) {}
    protected open fun onLoginFailCallback(resultCode: Int, bundle: Bundle?) {}
    override fun onDestroy() {
        super.onDestroy()
        if (loginLauncher != null) {
            loginLauncher!!.unregister()
        }
    }

    /**
     * 显示加载弹框
     *
     * @param dialogMessage 弹框内容，如果内容为空则不展示文字部分
     */
    private fun showLoadingDialog(dialogMessage: String, isCanCancel: Boolean) {
        CustomProgressDialog.getInstance(context)
            .setCanCancel(isCanCancel)
            .setMessage(dialogMessage)
            .builder()
            .show()
    }

    override fun showLoading(dialogMessage: String) {
        requireActivity().runOnUiThread { showLoadingDialog(dialogMessage, false) }
    }

    override fun hideLoading() {
        requireActivity().runOnUiThread { CustomProgressDialog.getInstance(context).dismiss() }
    }

    override fun refreshLoading(dialogMessage: String) {
        requireActivity().runOnUiThread  {
            CustomProgressDialog.getInstance(requireActivity())
                .refreshMessage(dialogMessage)
        }
    }

    override fun showToast(msg: String) {
        requireActivity().runOnUiThread { ToastUtils.showShort(activity, msg) }
    }

    override fun onErrorCode(model: BaseModelEntity<*>?) {
        if (model == null) {
            return
        }
        if (!errorService.isLogin(model.code)) {
            errorService.toLogin(requireContext(), loginLauncher)
            return
        }
        if (!errorService.hasPermission(model.code)) {
            errorService.toNoPermission(requireContext())
        }
    }

    @JvmOverloads
    fun startActivity(toClx: Class<*>?, bundle: Bundle? = null) {
        val intent = Intent(requireContext(), toClx)
        if (bundle != null) {
            intent.putExtras(bundle)
        }
        startActivity(intent)
    }

    @JvmOverloads
    fun startForResult(
        activityResultLauncher: ActivityResultLauncher<Intent?>,
        toClx: Class<*>?,
        bundle: Bundle? = null
    ) {
        val intent = Intent(requireContext(), toClx)
        if (bundle != null) {
            intent.putExtras(bundle)
        }
        activityResultLauncher.launch(intent)
    }

    fun startForResult(activityResultLauncher: ActivityResultLauncher<Intent?>, intent: Intent?) {
        activityResultLauncher.launch(intent)
    }

}